### PR TITLE
kvstore.row_sparse_pull for GPU and end-to-end benchmark: CPU vs. multi-GPUs

### PR DESCRIPTION
**Pull request by reminisce**
# Overview
1. Extended kvstore.row_sparse_pull to support output arrays on GPUs and added gpu unit test.
2. Added benchmark script for sparse end-to-end flow (modified base on [example/sparse/linear_regression.py](https://github.com/eric-haibin-lin/mxnet/blob/sparse/example/sparse/linear_regression.py)).

@piiswrong @eric-haibin-lin @madjam @stefanhenneking @anirudh2290 @mli

# Bencmark Results
**Hardware**: p2.8xlarge
**Model**:
```python
x = mx.symbol.Variable("data", stype='csr')
norm_init = mx.initializer.Normal(sigma=0.01)
w = mx.symbol.Variable("w", shape=(feature_dim, args.output_dim), init=norm_init, stype='row_sparse')
embed = mx.symbol.dot(x, w)
y = mx.symbol.Variable("softmax_label")
model = mx.symbol.SoftmaxOutput(data=embed, label=y, name="out")
```

**Note**: The CPU benchmark used the MXNet lib compiled **without** `USE_CUDA=1`.
### avazu dataset
**Feature Dim Size**: 1,000,000

| device |  batch size | output dim size | runtime (s) / epoch | device |  batch size | output dim size | runtime (s) / epoch |
| :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| cpu/32 cores | 512 * 1 | 512 | 196 | gpu/1 card | 512 * 64 * 1 | 512 | 24 |
| cpu/32 cores | 512 * 2 | 512 | 109 | gpu/2 cards | 512 * 64 * 2 | 512 | 20 |
| cpu/32 cores | 512 * 4 | 512 | 63 | gpu/3 cards | 512 * 64 * 3 | 512 | 20 |
| cpu/32 cores | 512 * 8 | 512 | 42 | gpu/4 cards | 512 * 64 * 4 | 512 | 20 |
| cpu/32 cores | 512 * 16 | 512 | 30 | gpu/5 cards | 512 * 64 * 5 | 512 | 20 |
| cpu/32 cores | 512 * 32 | 512 | 25 | gpu/6 cards | 512 * 64 * 6 | 512 | 20 |
| cpu/32 cores | 512 * 64 | 512 | 23 | gpu/7 cards | 512 * 64 * 7 | 512 | 20 |
| cpu/32 cores | 512 * 128 | 512 | 22 | gpu/8 cards | 512 * 64 * 8 | 512 | 21 |
| cpu/32 cores | 512 * 1 | 1,024 | 359 | gpu/1 card | 512 * 64 * 1 * 64 | 1,024 | 49 |
| cpu/32 cores | 512 * 2 | 1,024 | 198 | gpu/2 cards | 512 * 64 * 2 * 64 | 1,024 | 41 |
| cpu/32 cores | 512 * 4 | 1,024 | 121 | gpu/3 cards | 512 * 64 * 3 | 1,024 | 39 |
| cpu/32 cores | 512 * 8 | 1,024 | 79 | gpu/4 cards | 512 * 64 * 4| 1,024 | 39 |
| cpu/32 cores | 512 * 16 | 1,024 | 58 | gpu/5 cards | 512 * 64 * 5 | 1,024 | 38 |
| cpu/32 cores | 512 * 32 | 1,024 | 50 | gpu/6 cards | 512 * 64 * 6 | 1,024 | 40 |
| cpu/32 cores | 512 * 64 | 1,024 | 46 | gpu/7 cards | 512 * 64 * 7 | 1,024 | 40 |
| cpu/32 cores | 512 * 128 | 1,024 | 44  | gpu/8 cards | 512 * 64 * 8 | 1,024 | 41 |

### kdda dataset
**Feature Dim Size**: 20,216,830

| device |  batch size | output dim size | runtime (s) / epoch | device |  batch size | output dim size | runtime (s) / epoch |
| :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| cpu/32 cores | 512 * 1 | 32 | 123 | gpu/1 card | 512 * 64 * 1 | 32 | 20 |
| cpu/32 cores | 512 * 2 | 32 | 71 | gpu/2 cards | 512 * 64 *2 | 32 | 20 |
| cpu/32 cores | 512 * 4 | 32 | 46 | gpu/3 cards | 512 * 64 * 3 | 32 | 20 |
| cpu/32 cores | 512 * 8 | 32 | 33 | gpu/4 cards | 512 * 64 * 4| 32 | 20 |
| cpu/32 cores | 512 * 16 | 32 | 27 | gpu/5 cards | 512 * 64 * 5 | 32 | 20 |
| cpu/32 cores | 512 * 32 | 32 | 24 | gpu/6 cards | 512 * 64 * 6 | 32 | 20 |
| cpu/32 cores | 512 * 64 | 32 | 22 | gpu/7 cards | 512 * 64 * 7 | 32 | 20 |
| cpu/32 cores | 512 * 128 | 32 | 21 | gpu/8 cards | 512 * 64 * 8 | 32 | 21 |
| cpu/32 cores | 512 * 1 | 64 | 200 | gpu/1 card | 512 * 64 * 1 | 64 | 40 |
| cpu/32 cores | 512 * 2 | 64 | 120 | gpu/2 cards | 512 * 64 * 2 | 64 | 40 |
| cpu/32 cores | 512 * 4 | 64 | 80 | gpu/3 cards | 512 * 64 * 3 | 64 | 40 |
| cpu/32 cores | 512 * 8 | 64 | 61 | gpu/4 cards | 512 * 64 * 4| 64 | 40 |
| cpu/32 cores | 512 * 16 | 64 | 50 | gpu/5 cards | 512 * 64 * 5 | 64 | 40 |
| cpu/32 cores | 512 * 32 | 64 | 45 | gpu/6 cards | 512 * 64 * 6 | 64 | 40 |
| cpu/32 cores | 512 * 64 | 64 | 43 | gpu/7 cards | 512 * 64 * 7 | 64 | 40 |
| cpu/32 cores | 512 * 128 | 64 | 42 | gpu/8 cards | 512 * 64 * 8 | 64 | 41 |

# Profiling Results
**avazu dataset, batch_size=512, output_dim=1024**

**Profiling 32-core CPU training**
image

**Profiling single GPU training**
image

**Profiling 8-GPU training** Run with env var `MXNET_CPU_WORKER_NTHREADS=8` to parallelize execution of 8 sparse_retain operations.
image

### CODE REVIEW

File: src/kvstore/comm.h [hunk 1]
```diff
@@ -170,34 +170,36 @@ class CommCPU : public Comm {
...
   void BroadcastRowSparse(int key, const NDArray& src,
                           const std::vector<std::pair<NDArray*, NDArray>>& dst,
                           const bool use_copy,
                           const int priority) override {
     using namespace mshadow;
-    auto size = dst.size();
-    for (size_t i = 0; i < size; i++) {
-      auto out = dst[i].first;
-      auto row_id = dst[i].second;
+    for (size_t i = 0; i < dst.size(); ++i) {
```
Comment (cjolivier01) [inline src/kvstore/comm.h]: nit: can be faster to not call size() every iteration, such as:
for(size_t i = 0, n = dst.size(); i < n; ++i)
I used to think that compilers always optimized these little calls out, but often they don't always do it, I found when using profilers

File: src/kvstore/comm.h [hunk 2]
```diff
@@ -170,34 +170,36 @@ class CommCPU : public Comm {
...
                  << "BroadcastRowSparse expects row_sparse dst NDArray";
-        CHECK_EQ(out->ctx().dev_mask(), Context::kCPU)
-                 << "BroadcastRowSparse with dst on gpu context not supported";
         CHECK_EQ(row_id.ctx().dev_mask(), Context::kCPU)
                  << "BroadcastRowSparse with src on gpu context not supported";
         // retain according to unique indices
-        Engine::Get()->PushSync([src, out, row_id](RunContext rctx) {
-            NDArray *output = out;
-            const auto indices = row_id.data();
+        bool is_to_gpu = out->ctx().dev_mask() == Context::kGPU;
```
Comment (cjolivier01) [inline src/kvstore/comm.h]: nit: const?

**Comment by eric-haibin-lin** [comment]
Since there's still some changes required to run multi-device example, should we separate the changes to `src/operator/tensor/square_sum-inl.h` in another PR and merge it first? @reminisce

File: src/kvstore/comm.h (hunk 1 above)
Comment (reminisce) [inline src/kvstore/comm.h]: I was wondering how much time could be saved using cached size and how large the size should be in order to see the difference? What is the compiler version?

Comment (cjolivier01) [inline src/kvstore/comm.h]: Function call overhead is relatively high if it doesn't get inlined for some reason. You probably wouldn't notice it in one pass, but if this function gets called millions+ of times during training, it can add up. It probably wouldn't be a bottleneck, but a CPU tick here and a CPU tick there eventually add up to real time. The question is why spend ticks when it's very simple not to?

Comment (cjolivier01) [inline src/kvstore/comm.h]: Also, even with inlining, you're less likely to get a memory cache miss when the value is on the stack...

Comment (reminisce) [inline src/kvstore/comm.h]: In most std implementations, vector::size is inlined. In terms of micro-optimization vs. readability and maintainability, I prefer the latter one unless that kind of optimization generates noticeable improvements. Falling into the habit of caching vector size is dangerous in practice. In addition, creating a local variable doesn't guarantee a performance improvement. If your register is full, reading the local variable from stack also has extra overhead. I agree that we may evaluate this kind of micro-optimization in the cases of dealing tight loops. However, the current loop is not trivial.

I also tried two approaches using the benchmark script in this PR and didn't find any noticeable difference between them.

Here are some good discussions:
https://stackoverflow.com/a/2753637
https://stackoverflow.com/a/8613000

File: src/kvstore/comm.h (hunk 2 above)
Comment (reminisce) [inline src/kvstore/comm.h]: Ok.

**Comment by reminisce** [comment]
@eric-haibin-lin I will submit the change in square_sum-inl.h in another PR.

**Comment by reminisce** [comment]
@eric-haibin-lin 8-GPU training with `MXNET_CPU_WORKER_NTHREADS=8`.
[profile_avazu_gpu8_nworker1_batchsize512_outdim512.json.zip](https://github.com/eric-haibin-lin/mxnet/files/1219243/profile_avazu_gpu8_nworker1_batchsize512_outdim512.json.zip)

**Comment by reminisce** [comment]
@eric-haibin-lin 8-GPU training with MXNET_CPU_WORKER_NTHREADS=8 after changing the locations of row_sparse_pull.
[profile_avazu_gpu8_nworker1_batchsize512_outdim512.json.zip](https://github.com/eric-haibin-lin/mxnet/files/1219280/profile_avazu_gpu8_nworker1_batchsize512_outdim512.json.zip)

**Comment by reminisce** [comment]
@eric-haibin-lin 8-GPU training with MXNET_CPU_WORKER_NTHREADS=8 kdda, after changing the locations of row_sparse_pull.
[profile_kdda_gpu8_nworker1_batchsize512_outdim32.json.zip](https://github.com/eric-haibin-lin/mxnet/files/1219330/profile_kdda_gpu8_nworker1_batchsize512_outdim32.json.zip)

**Comment by reminisce** [comment]
@eric-haibin-lin Profiling results after skipping sparse_retain to copying retained rows from cpu to gpu directly.
[profile_avazu_gpu1-8_batchsize32768_outputdim512.zip](https://github.com/eric-haibin-lin/mxnet/files/1221002/profile_avazu_gpu1-8_batchsize32768_outputdim512.zip)

**Comment by reminisce** [comment]
@eric-haibin-lin 8GPU after changing metric from MSE to Accuracy.
[profile_avazu_gpu8_nworker1_batchsize32768_outdim512.json.zip](https://github.com/eric-haibin-lin/mxnet/files/1221348/profile_avazu_gpu8_nworker1_batchsize32768_outdim512.json.zip)

File: src/executor/graph_executor.cc [hunk 3]
```diff
@@ -1151,7 +1151,9 @@ void GraphExecutor::InitDataEntryMemory(std::vector<NDArray>* shared_pool) {
       CHECK_LE(nword, std::numeric_limits<nnvm::dim_t>::max());
       // allocate float arrays
       TShape shape{static_cast<nnvm::dim_t>(nword)};
-      NDArray nd(shape, ctx);
+      // TODO(junwu): adding delay_alloc=true to create nd
+      // is a temporary solution.
+      NDArray nd(shape, ctx, true);
```
Comment (eric-haibin-lin) [inline src/executor/graph_executor.cc]: Adding `delay_alloc=True` should work just fine. Did it break any unit test?

File: benchmark/python/sparse_end2end.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+    # the weights to each context
+    # column indices (NDArray type) of the csr data
+    # used as the row_idx of the weight row-sparse matrix
+    # TODO(junwu):
+    # the following two lines block, may need to precompute
+    # them and cache them outside the for loop
+    row_indices = data.indices
+    indptr = data.indptr.asnumpy()
+    row_idx_array = []
+    for s in slices:
```
Comment (eric-haibin-lin) [inline benchmark/python/sparse_end2end.py]: We will reuse this script for multi-machien cpu training. we can skip slicing and `asnumpy()` call if `len(slices) == 1`.

File: benchmark/python/sparse_end2end.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+    # module
+    mod = mx.mod.Module(symbol=model, data_names=['data'],
+                        label_names=['softmax_label'], context=contexts)
+    mod.bind(data_shapes=train_data.provide_data, label_shapes=train_data.provide_label)
+    mod.init_params(initializer=mx.init.Uniform(scale=.1))
+    sgd = mx.optimizer.SGD(momentum=0.0, clip_gradient=5.0,
+                           learning_rate=0.1, rescale_grad=1.0/batch_size/num_worker)
+    mod.init_optimizer(optimizer=sgd, kvstore=kv)
+    # use accuracy as the metric
+    metric = mx.metric.create('MSE')
```
Comment (eric-haibin-lin) [inline benchmark/python/sparse_end2end.py]: Just change to `Accuracy ` for benchmarking is fine

Comment (reminisce) [inline benchmark/python/sparse_end2end.py]: Done.

File: benchmark/python/sparse_end2end.py (hunk 4 above)
Comment (reminisce) [inline benchmark/python/sparse_end2end.py]: Done.

File: src/executor/graph_executor.cc (hunk 3 above)
Comment (reminisce) [inline src/executor/graph_executor.cc]: No, it doesn't break any unit tests.

**Comment by reminisce** [comment]
Final profiling results for 6-GPU training dataset avazu, `output_dim=512, batch_size=32768, MXNET_CPU_WORKER_NTHREADS=6`. Replaced `update_metric` with `mx.nd.waitall()` since `update_metric` has too many sync copy from GPU to CPU that can be optimized out. Total 6-GPU training runtime: 17.8 seconds, while 1-GPU training costs 24 seconds.

image

image
